iree/compiler/Translation/HALExecutable.cpp

```diff
@@ -0,0 +1,82 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Translation/HALExecutable.h"
+
+#include "iree/compiler/Dialect/HAL/Transforms/Passes.h"
+#include "iree/compiler/Utils/TracingUtils.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Translation.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+// Converts from a module containing a hal.executable into the serialized form.
+static LogicalResult translateFromMLIRToHALExecutable(
+    mlir::ModuleOp moduleOp, IREE::HAL::TargetOptions executableOptions) {
+  auto executableOps =
+      llvm::to_vector<4>(moduleOp.getOps<IREE::HAL::ExecutableOp>());
+  auto sourceOps =
+      llvm::to_vector<4>(moduleOp.getOps<IREE::HAL::ExecutableSourceOp>());
+  size_t usableOpCount = executableOps.size() + sourceOps.size();
+  if (usableOpCount != 1) {
+    return moduleOp.emitError()
+           << "HAL executable translation requires "
+              "exactly 1 top level hal.executable/hal.executable.source op";
```

Comment:
Thanks for the error message. I hit this immediately when I tried using my original test file instead of one with `hal.executable.source` ops :P